File: basic-tut-3.py
#!/usr/bin/env python3
import sys
import gi
import logging
import signal

gi.require_version("GLib", "2.0")
gi.require_version("GObject", "2.0")
gi.require_version("Gst", "1.0")
from gi.repository import GLib, GObject, Gst

logging.basicConfig(
    level=logging.DEBUG, format="[%(name)s] [%(levelname)8s] - %(message)s"
)
logger = logging.getLogger(__name__)


class CustomData:
    def __init__(self):
        # Create Elements
        # uridecodebin -> Decodes data from a URI into raw media (instantiates source, demuxer, decoder internally)
        self.source = Gst.ElementFactory.make("uridecodebin", "source")
        # audioconvert -> Convert audio to different formats. Ensures platform interoperability
        self.convert = Gst.ElementFactory.make("audioconvert", "convert")
        # audioresample -> Useful for converting between different audio sample rates. Again ensures platform interoperability
        self.resample = Gst.ElementFactory.make("audioresample", "resample")
        # autoaudiosink -> render the audio stream to the audio card
        self.sink = Gst.ElementFactory.make("autoaudiosink", "sink")
        # Create Empty pipeline
        self.pipeline = Gst.Pipeline.new("test-pipeline")


# callback function
# src is the GstElement which triggered the signal.
# new_pad is the GstPad that has just been added to the src element
# This is the pad to which we want to link.


def pad_added_handler(src, new_pad, data):
    sink_pad = data.convert.get_static_pad("sink")
    print(f"Received new pad {new_pad.get_name()} from {src.get_name()}")

    if sink_pad.is_linked():
        print("Already linked")
        return

    new_pad_caps = new_pad.get_current_caps()
    new_pad_struct = new_pad_caps.get_structure(0)
    new_pad_type = new_pad_struct.get_name()

    if not new_pad_type.startswith("audio/x-raw"):
        print("Type not audio.")
        return

    # Linking
    ret = new_pad.link(sink_pad)
    if ret == Gst.PadLinkReturn.OK:
        print("Link success")
    else:
        print("Link failure")

    return


def main():
    # Initialize GStreamer library
    Gst.init(None)
    signal.signal(signal.SIGINT, signal.SIG_DFL)  # signal.signal(signalnum, handler)
    data = CustomData()
    if (
        not data.pipeline
        or not data.source
        or not data.convert
        or not data.resample
        or not data.sink
    ):
        logger.error("Not all elements could be created")
        sys.exit(1)

    # Add elements to pipeline, the elements are not linked yet!
    data.pipeline.add(data.source, data.convert, data.resample, data.sink)

    # Link all elements excluding source (which contains demux, hence dynamic)
    ret = data.convert.link(data.resample)
    ret = ret and data.resample.link(data.sink)
    if not ret:
        logger.error("Elements could not be linked")
        sys.exit(1)

    # Set URI to play
    data.source.props.uri = "https://www.freedesktop.org/software/gstreamer-sdk/data/media/sintel_trailer-480p.webm"

    # The uridecodebin element comes with several element signals including `pad-added`
    # When uridecodebin(source) creates a source pad, and emits `pad-added` signal, the callback is invoked
    # Non-blocking
    data.source.connect("pad-added", pad_added_handler, data)

    ret = data.pipeline.set_state(Gst.State.PLAYING)
    if ret == Gst.StateChangeReturn.FAILURE:
        logger.error("Unable to change state of pipeline to playing")
        sys.exit(1)

    # Listening for messages: Wait for EOS or error
    bus = data.pipeline.get_bus()
    while True:
        msg = bus.timed_pop_filtered(
            Gst.CLOCK_TIME_NONE,
            Gst.MessageType.ERROR | Gst.MessageType.STATE_CHANGED | Gst.MessageType.EOS,
        )
        # Parse message
        if msg:
            if msg.type == Gst.MessageType.ERROR:
                err, debug_info = msg.parse_error()
                logger.error(
                    f"Error received from element {msg.src.get_name()}: {err.message}"
                )
                logger.error(
                    f"Debugging information: {debug_info if debug_info else 'none'}"
                )
                break
            elif msg.type == Gst.MessageType.EOS:
                logger.info("End-Of-Stream reachedata.")
                break

            elif msg.type == Gst.MessageType.STATE_CHANGED:
                if msg.src == data.pipeline:
                    old_state, new_state, pending_state = msg.parse_state_changed()
                    print(
                        f"Pipeline state changed from {Gst.Element.state_get_name(old_state)} to {Gst.Element.state_get_name(new_state)}"
                    )
            else:
                logger.error("Unexpected message")
                break


# Main section
if __name__ == "__main__":
    main()

File: C-code/basic-tut3.c
#include <gst/gst.h>
#include <stdio.h>

typedef struct CustomData{
  GstElement *source;
  GstElement *convert;
  GstElement *resample;
  GstElement *sink;
  GstElement *pipeline;
} CustomData;

void pad_added_handler(GstElement *src, GstPad *newpad, CustomData *data){
  GstPad *sink_pad = gst_element_get_static_pad(data -> convert, "sink");
  printf("Received new pad \n");
  if(gst_pad_is_linked(sink_pad)){
    fprintf(stderr, "Already linked");
  }
  GstCaps *newpad_caps = gst_pad_get_current_caps(sink_pad);
  GstStructure *newpad_struct = gst_caps_get_structure(newpad_caps, 0);
  const gchar *newpad_type = gst_structure_get_name(newpad_struct);
  if (!g_str_has_prefix (newpad_type, "audio/x-raw")) {
    g_print ("Ignoring. Expected type: Raw Audio. Actual type: '%s' \n", newpad_type);
  }

  if (gst_pad_link(newpad, sink_pad) != 0 ){
    fprintf(stderr, "Linking failed \n");
  }
  else {
    fprintf(stderr, "Link success \n");
  }
}

int main(int argc, char * argv[]){
  CustomData data;
  GstBus *bus;
  GstMessage *msg;
  gboolean terminate = FALSE;
  gst_init(&argc, &argv);
  data.source = gst_element_factory_make("uridecodebin", "source");
  data.convert = gst_element_factory_make("audioconvert", "convert");
  data.resample = gst_element_factory_make("audioresample", "resample");
  data.sink = gst_element_factory_make("autoaudiosink", "sink");
  data.pipeline = gst_pipeline_new("test-pipeline");

  if(!data.source || !data.convert || !data.resample || !data.sink || !data.pipeline){
    fprintf(stderr, "Not all elements could be created.\n");
    return -1;
  }

  gst_bin_add_many(GST_BIN(data.pipeline), data.source, data.convert, data.resample, data.sink, NULL);

  if (!gst_element_link_many(data.convert, data.resample, data.sink, NULL)){
    fprintf(stderr, "Not all elements could be linked.\n");
    return -1;
  }
  // Set URI to play
  g_object_set (data.source, "uri", "https://www.freedesktop.org/software/gstreamer-sdk/data/media/sintel_trailer-480p.webm", NULL);

  // Connect to the pad-added handler
  g_signal_connect(data.source, "pad-added", G_CALLBACK(pad_added_handler), &data);

// move tpipeline to playing state
  int ret = gst_element_set_state(data.pipeline, GST_STATE_PLAYING);
  if (ret == 0){
    fprintf(stderr, "Could not change pipeline state to playing\n");
    return -1;
  }
  bus = gst_element_get_bus(data.pipeline);
  do {
    msg = gst_bus_timed_pop_filtered(bus, GST_CLOCK_TIME_NONE, GST_MESSAGE_EOS | GST_MESSAGE_ERROR | GST_MESSAGE_STATE_CHANGED);
    GError *err = NULL;
    gchar *dbg_info = NULL;
    switch(GST_MESSAGE_TYPE(msg)){
      case GST_MESSAGE_EOS:
        printf("End of stream reached.\n");
        terminate = TRUE;
        break;
      case GST_MESSAGE_ERROR:
        gst_message_parse_error(msg, &err, &dbg_info);
        g_print("Error received: %s \n", err->message);
        g_print("Error from: %s\n", GST_OBJECT_NAME(msg->src));
        g_print("Debug info: %s\n", dbg_info);
        terminate = TRUE;
        break;

      case GST_MESSAGE_STATE_CHANGED:
        if (GST_MESSAGE_SRC(msg) == GST_OBJECT(data.pipeline)){
          GstState old_state, new_state;
          gst_message_parse_state_changed(msg, &old_state, &new_state, NULL);
          g_print("Pipeline state has changed from %s to %s\n", gst_element_state_get_name(old_state), gst_element_state_get_name(new_state));
        }
        break;
      default:
        g_print("Unexpected message received \n");
        break;
  }
} while(!terminate);

  gst_message_unref(msg);
  gst_object_unref(bus);
  gst_element_set_state(data.pipeline, GST_STATE_NULL);
  gst_object_unref(data.pipeline);

  return 0;
}
